Allow defining an EventDispatcher service in the container

weierophinney opened the pull request:
This patch adds logic to the `ApplicationFactory` to test for a `Laminas\Cli\SymfonyEventDispatcher` instance and, if found, use it to populate the symfony/console instance (if not found, we instantiate one directly, as we did before).

We **require** that the service returned implement `Symfony\Component\EventDispatcher\EventDispatcherInterface` to ensure that we can add our `ConsoleEvents::TERMINATE` event.

Users who want to supply additional listeners or subscribers can do so by defining the service and optionally adding delegators to it.

Fixes #47

weierophinney committed:
feat: allow providing a custom event dispatcher instance

This patch adds logic to the `ApplicationFactory` to test for a `Laminas\Cli\SymfonyEventDispatcher` instance and, if found, use it to populate the symfony/console instance (if not found, we instantiate one directly, as we did before).

We **require** that the service returned implement `Symfony\Component\EventDispatcher\EventDispatcherInterface` to ensure that we can add our `ConsoleEvents::TERMINATE` event.

Users who want to supply additional listeners or subscribers can do so by defining the service and optionally adding delegators to it.

Fixes #47

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Note: this patch does not contain the fixes found in #53, so tests will fail until that patch is merged.

weierophinney commented:
Pushed all QA changes into separate PRs that are now in the 0.1.x branch, which allows tests for this to run; ready to review now, @michalbundyra .

froschdesign commented on docs/book/events.md:
```diff
@@ -0,0 +1,44 @@
+# Events
+
+The symfony/console component allows attaching an event dispatcher instance to a console application.
```
```suggestion
The [symfony/console component](https://symfony.com/doc/current/components/console.html) allows attaching an event dispatcher instance to a console application.
```

froschdesign commented on docs/book/events.md:
````diff
@@ -0,0 +1,44 @@
+As an example, your container configuration file might look like the following for laminas-mvc applications:
+
+```php
+// in config/autoload/console.global.php
````
_Excerpt from the upcoming documentation writer's guide:_

### Filenames

A filename must be placed before a code example because there are always two parts for this information: 

1. what kind of file is this
2. and what is the name or path.

Both informations in a code comment would be too long and misplaced.

#### Examples

> Create a file for the module configuration called `module.config.php` under `laminas-mvc-tutorial/module/Album/config/`:

> …in a factory, e.g. `src/Album/Controller/AlbumControllerFactory.php`:

> …in a module configuration file, e.g. `module/Album/config/module.config.php`:

froschdesign commented on docs/book/events.md:
````diff
@@ -0,0 +1,44 @@
+Or like the following for Mezzio applications:
+
+```php
+// in config/autoload/console.global.php
````
The same here.

weierophinney committed:
docs: incorporate suggestions per review

- Link to symfony/console docs
- Introduce file and file type in narrative, instead of code sample

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra reviewed (commented):
@weierophinney it looks good, but I am missing something. I'd like to have either:

- Link to the documentation how to use events https://symfony.com/doc/current/components/console/events.html
or
- Full example in our documentation - with example event
(or both ...)

I'd like to see full example here, so it's clear what is our recommended way of adding listeners. Do we need to write a factory if we want to use default event listener and just attach some listeners? (or we can use invokable/alias... ?)
What people might want here is a simple solution to add custom listener, and this is the example I am missing.

Also what I just noted in the Symfony Docs is:

> Console events are only triggered by the main command being executed. Commands called by the main command will not trigger any event.

So I think it might be worth noting also on our side, that the events are not triggered for chained commands.

froschdesign commented:
> I'd like to see full example here, so it's clear what is our recommended way of adding listeners.

Comprehensible examples are a must! 👍

weierophinney committed:
docs: provide full example of wiring a listener

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@froschdesign and @michalbundyra — Updated the `events.md` to provide a full example!

weierophinney committed:
docs: link to symfony/console events documentation

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: consistency in code samples

Ensure an opening `<?php` line on each PHP example in the new doc.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
